### ISSUE TITLE
Incremental training

### ISSUE THREAD

**Issue by tabergma** [RasaHQ/rasa#6971]
**Description of Problem**:
Once a model is trained it cannot be updated. It is not possible to continue training the model on new data that came in. Instead the model needs to be retrained from scratch, which takes up a lot of time.

**Overview of the Solution**:
It should be possible to load a model from a previous checkpoint and continue training with new data added.

**Comment by evgeniiaraz** [comment]
@tabergma how urgent is this one?

**Comment by tabergma** [comment]
@evgeniiaraz We want to tackle this issue this quarter. @dakshvar22 is leading this topic. Why do you ask?

**Comment by evgeniiaraz** [comment]
@tabergma I wanted to work on it not to lose shape :)  but if it is urgent, I'll pick something easier

**Comment by dakshvar22** [comment]
Based on the discussion in the [document](https://docs.google.com/document/d/1B7iVPRQONPoFssJfxhLMm5o23KWTPYNPQw_7XRc38Ao/edit), here are more fine-grained implementation tasks - 

**Changes to CLI and `rasa/train.py`**
1. Add a parameter to `rasa train` called `finetune_previous_model` which starts training in fine-tuning mode.
2. Add a parameter to `rasa train` called `finetune_model_path` which lets you specify the path to a previous model which should be finetuned.
3. `rasa.train_async_internal` should be refactored to check if training should proceed in finetuning mode(if `finetune_previous_model` is set to True). If yes, then it should check if it is possible to do so.(Check the doc for constraints for finetuning to be possible).
4. `rasa.nlu.train` should be refactored to create the `Trainer` object in fine-tune mode which means each component should be loaded with the model to be finetuned from. This will involve building the pipeline in a way similar to how the pipeline is built during inference, i.e. when `rasa shell` is run or `rasa test` is run.
5. `rasa.core.train` should also be refactored similarly as above.

**Changes needed to ML components**

*CountVectorsFeaturizer*(CVF)

1. Add a parameter `max_additional_vocabulary_size` which lets users specify the additional buffer size that CVF should keep to accommodate new vocabulary tokens during fine-tuning.
2. `_train_with_independent_vocab` should be refactored to construct the vocabulary with the additional buffer specified above. Things to keep in mind here -

- When a new training cycle is triggered, the ordering of existing vocabulary tokens should not be changed and the new vocabulary tokens should only occupy the empty slots in the vocabulary.
- If the vocabulary size of CVF is exhausted, we should continue training, but warn the user that the vocabulary is exhausted and treat the new tokens which overflow as OOV tokens. At this point, the user should also be informed about the total vocabulary size of their dataset and should be prompted to retrain with full vocabulary.

*DIETClassifier, ResponseSelector and TEDPolicy*

1. `load()` should be refactored to load the models with weights in training mode and not in prediction mode. Currently, `_load_model()` builds the TF graph in predict mode which should be changed if the classifier is being loaded for finetuning. So instead of calling `_get_tf_call_model_function`, `_get_tf_train_functions` should be reused to build the graph for training.
2. Make sure the signature of `RasaModelData` in finetune mode is the same as what is constructed during training from scratch.

**Comment by dakshvar22** [comment]
A working version(very draft) of the above steps is implemented on this [branch](https://github.com/RasaHQ/rasa/tree/incremental_training). What needs to be refactored better and possible catch all edge cases are changes outside of the ML components.

**Comment by wochinge** [comment]
Next steps based on the call with @dakshvar22 @joejuzl 
1. Create engineering issues from this (should be around 2-3 issues 🤔 )
2. Get started with the engineering issues in the week of November 23rd

Other things to keep in mind:
* Can we branch off `master` or does it make more sense to branch off the `e2e` branch?

**Comment by dakshvar22** [comment]
I ran some initial experiments using the working version on this [branch](https://github.com/RasaHQ/rasa/tree/incremental_training) - 

**Setup**

Data: Financial Bot [NLU data](https://github.com/RasaHQ/financial-demo/blob/rasa-2-0/data/nlu/nlu.yml) split into 80:20 train test split. The train split is further divided into 2 sets - split 80:20. The first set is used for training an initial model from scratch. The second set is used for finetuning the first model trained. Consider the second set as new annotations that a user added to their training data.

Size of Set 1: 233
Size of Set 2: 292
Size of held-out test set: 73

*Note*: Finetuning is done by simply mixing the new data with the old data and then training on batches from the combined data.

Training: We train the first model from scratch for 100 epochs. Then add the second set to the training data and further train the first model for 30 more epochs.

[Config](https://github.com/RasaHQ/financial-demo/blob/rasa-2-0/config.yml)

Results:

**Initial Model**|**Training data**|**Number of epochs**|**Intent F1(held out test set)**|**Entity F1(held out test set)**|**Time for training**
:-----:|:-----:|:-----:|:-----:|:-----:|:-----:
Randomly initialized| Set 1| 100| 0.753| 0.9| 48s
Model trained on set 1| Set 1 + Set 2| 30| 0.861| 0.927| 16s
Randomly initialized| Set 1 + Set 2| 130| 0.876| 0.911| 1 min 16s

**Comment by dakshvar22** [comment]
Experiments on Sara data - 

Size of Set 1: 3166
Size of Set 2: 792
Size of held-out test set: 990

Results:

**Initial Model**|**Training data**|**Number of epochs**|**Intent F1**|**Entity F1**|**Response F1**|**Time for training**
:-----:|:-----:|:-----:|:-----:|:-----:|:-----:|:-----:
Randomly initialized| Set 1| 40| 0.789| 0.832| 0.927| 4m 10s
Model trained on set 1| Set 1 + Set 2| 10| 0.823| 0.861| 0.935| 1m 39s
Randomly initialized| Set 1 + Set 2| 50| 0.818| 0.854| 0.938| 6m 2s

**Comment by wochinge** [comment]
@dakshvar22 Do I understand it correctly that the incremental training is in total faster than training everything at once? This seems somewhat counterintuitive for me as I'd expect overhead from loading training data / pipelines etc.

**Comment by dakshvar22** [comment]
@wochinge The time mentioned above are the time to train `DIETClassifier` alone and does not include the pipeline and training data loading time. We should measure that too, but it would be much smaller in comparison to the amount of time required to train `DIETClassifier` for an additional 40/70 epochs as shown in the examples above.

**Comment by wochinge** [comment]
Thanks for clarifying! Even if we measure the `DIETClassier` on its own - shouldn't the total time of the incremental timing be greater than when training everything in one go?

**Comment by dakshvar22** [comment]
@wochinge The small overhead(11s) that you see when trained in one go is because of the increase in input feature vector size and hence bigger matrix multiplications. The first two experiments on Sara data have an input feature vector of size 11752(actual vocabulary size + buffer added). The third experiment has an input feature vector of size 12752(actual vocabulary size + buffer added). The additional 1000 dimensions are present because the model is trained from scratch and hence new buffer space is added in `CountVectorsFeaturizer`. I did run an additional experiment to validate this with `additional_vocabulary_size` set to 0 in `CountVectorsFeaturizer` and the training times were then comparable with a small stochastic overhead(+-2 secs) either side. Does that help clarify?

**Comment by wochinge** [comment]
Thanks a lot for digging into and clarifying this! 🙌

**Comment by wochinge** [comment]
I had a short look on the `e2e` branch and at least for the engineering changes we don't need to branch off [e2e](https://github.com/RasaHQ/rasa/compare/e2e). However, `DietClassifier` has huge changes @dakshvar22  so you want to probably branch off from `e2e` for your changes, what do you think?

**Comment by dakshvar22** [comment]
@wochinge The only change that we need for incremental training inside `DIETClassifier` is a change in the `load` method which isn't touched on `e2e`. So, we should be fine branching off master. Would like to decouple it from e2e as much as possible.

**Comment by dakshvar22** [comment]
@wochinge @joejuzl Created a shared branch named [`continuous_training`](https://github.com/RasaHQ/rasa/tree/continuous_training) for us to merge our respective PRs into.

**Comment by wochinge** [comment]
@dakshvar22 cc @joejuzl Can we finetune a core model when NLU was finetuned previously? Or do we have to train Core from scratch as the featurization of messages will change?

**Comment by dakshvar22** [comment]
Not sure if I understand the case completely. Do you mean that `rasa train nlu finetune` was run and then `rasa train core finetune` was run?

**Comment by wochinge** [comment]
1. We run `rasa train --finetune`
2. NLU model is finetuned
3. Do we now finetune the core model or do we train it from scratch?

**Comment by dakshvar22** [comment]
Ohh, we can finetune the core model as long as we are inside our current constraints, i.e. no change to labels(intents, actions, slots, entities, etc.). Why do you think we would need to train it from scratch?
